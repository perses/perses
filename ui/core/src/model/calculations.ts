// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import findLast from 'lodash/findLast';
import meanBy from 'lodash/meanBy';
import sumBy from 'lodash/sumBy';
import { TimeSeriesValueTuple } from '@perses-dev/core';

// TODO: move this file and calculations.test.ts to @perses-dev/core
export const CalculationsMap = {
  First: first,
  Last: last,
  LastNumber: lastNumber,
  Mean: mean,
  Sum: sum,
};

export type CalculationType = keyof typeof CalculationsMap;

export type CalculationConfig = {
  label: string;
  description: string;
};
export const CALCULATIONS_CONFIG: Readonly<Record<CalculationType, CalculationConfig>> = {
  First: {
    label: 'First',
    description: 'First value',
  },
  Last: {
    label: 'Last',
    description: 'Last value',
  },
  LastNumber: {
    label: 'Last number',
    description: 'Last numeric value',
  },
  Mean: {
    label: 'Mean',
    description: 'Average value',
  },
  Sum: {
    label: 'Sum',
    description: 'The sum of all values',
  },
} as const;

export const DEFAULT_CALCULATION: CalculationType = 'Sum';

function first(values: TimeSeriesValueTuple[]): number | undefined {
  const tuple = values[0];
  return tuple === undefined ? undefined : getValue(tuple);
}

function last(values: TimeSeriesValueTuple[]): number | undefined {
  if (values.length <= 0) return undefined;

  const tuple = values[values.length - 1];
  return tuple === undefined ? undefined : getValue(tuple);
}

function lastNumber(values: TimeSeriesValueTuple[]): number | undefined {
  const tuple = findLast(values, (tuple) => isNaN(getValue(tuple)) === false);
  return tuple === undefined ? undefined : getValue(tuple);
}

function mean(values: TimeSeriesValueTuple[]): number | undefined {
  if (values.length <= 0) return undefined;
  return meanBy(values, getValue);
}

function sum(values: TimeSeriesValueTuple[]): number | undefined {
  if (values.length <= 0) return undefined;
  return sumBy(values, getValue);
}

function getValue(valueTuple: TimeSeriesValueTuple) {
  const value = valueTuple[1];
  if (value !== null) {
    return value;
  }
  // TODO: refactor utils so null can be returned and LastNotNull supported
  return NaN;
}
