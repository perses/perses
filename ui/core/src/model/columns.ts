export interface ColumnOptions {
  name: string;

  // Text to display in the header for the column.
  header?: string;
  /**
   * Text to display when hovering over the header text. This can be useful for
   * providing additional information about the column when you want to keep the
   * header text relatively short to manage the column width.
   */
  headerDescription?: string;
  /**
   * Text to display when hovering over a cell. This can be useful for
   * providing additional information about the column when the content is
   * ellipsized to fit in the space.
   */
  cellDescription?: string;
  // Alignment of the content in the cell.
  align?: 'left' | 'center' | 'right';

  // When `true`, the column will be sortable.
  enableSorting?: boolean;
  /**
   * Width of the column when rendered in a table. It should be a number in pixels
   * or "auto" to allow the table to automatically adjust the width to fill
   * space.
   */
  width?: number | 'auto';
  // When `true`, the column will not be displayed.
  hide?: boolean;
}
