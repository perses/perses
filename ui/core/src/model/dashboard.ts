// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { DatasourceSpec } from './datasource';
import { LayoutDefinition } from './layout';
import { PanelDefinition } from './panels';
import { ProjectMetadata } from './resource';
import { DurationString } from './time';
import { VariableDefinition } from './variables';
import { Display } from './display';

export interface DashboardResource {
  kind: 'Dashboard';
  metadata: ProjectMetadata;
  spec: DashboardSpec;
}

export interface DashboardSpec {
  display?: Display;
  datasources?: Record<string, DatasourceSpec>;
  duration: DurationString;
  variables: VariableDefinition[];
  layouts: LayoutDefinition[];
  panels: Record<string, PanelDefinition>;
}

export interface DashboardSelector {
  project: string;
  dashboard: string;
}
