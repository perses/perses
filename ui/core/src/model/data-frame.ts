// TODO: Revisit whether something like danfo.js is appropriate to use here once
// they've finished their Typescript rewrite (https://danfo.jsdata.org/)
export interface DataFrame {
  name: string;
  columns: Series[];
}

type SeriesMap = {
  Date: UnixTimeMs;
  String: string;
  Number: number;
};

export type UnixTimeMs = number;

export type SeriesType = keyof SeriesMap;

export interface Series<T extends SeriesType = SeriesType> {
  seriesType: T;
  name: string;
  values: Array<SeriesMap[T]>;
}
