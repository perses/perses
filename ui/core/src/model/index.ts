// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export * from './calculations';
export * from './dashboard';
export * from './datasource';
export * from './definitions';
export * from './display';
export * from './ephemeraldashboard';
export * from './http';
export * from './http-proxy';
export * from './kind';
export * from './layout';
export * from './legend';
export * from './notice';
export * from './panels';
export * from './project';
export * from './query';
export * from './resource';
export * from './roles';
export * from './rolebindings';
export * from './secrets';
export * from './thresholds';
export * from './time';
export * from './time-series-data';
export * from './time-series-queries';
export * from './trace-data';
export * from './units';
export * from './user';
export * from './variables';
