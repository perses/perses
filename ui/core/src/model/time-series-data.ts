// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Notice } from './notice';
import { AbsoluteTimeRange } from './time';
import { Labels, TimeSeriesValueTuple } from './time-series-queries';

export interface TimeScale {
  startMs: number;
  endMs: number;
  stepMs: number;
  rangeMs: number;
}

export interface TimeSeriesData {
  timeRange?: AbsoluteTimeRange;
  stepMs?: number;
  series: TimeSeries[];
  metadata?: TimeSeriesMetadata;
}

export interface TimeSeries {
  name: string;
  values: TimeSeriesValueTuple[];
  histograms?: any[]; // eslint-disable-line @typescript-eslint/no-explicit-any
  formattedName?: string;
  labels?: Labels;
}

export interface TimeSeriesMetadata {
  notices?: Notice[];

  /**
   * The raw query that is executed to generate this data.
   * Useful when needing to inspect the query that was executed
   * after variables and other context modifications have been applied.
   */
  executedQueryString?: string;
}
