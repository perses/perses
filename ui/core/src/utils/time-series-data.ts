// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { AbsoluteTimeRange, TimeScale, TimeSeries, TimeSeriesData, TimeSeriesValueTuple } from '../model';
import { gcd } from './mathjs';

export const MIN_STEP_INTERVAL_MS = 10;

/**
 * Given a common time scale (see `getCommonTimeScale`), generates an array of
 * timestamp values in ms for the x axis of a graph.
 */
export function getXValues(timeScale: TimeScale): number[] {
  const xValues: number[] = [];
  let timestamp = timeScale.startMs;
  while (timestamp <= timeScale.endMs) {
    xValues.push(timestamp);
    timestamp += timeScale.stepMs;
  }
  return xValues;
}

/**
 * [DEPRECATED] Used for legacy LineChart 'category' axis approach.
 * Given a TimeSeries from a query and a common time scale (see `getCommonTimeScale`),
 * processes the time series values, filling in any timestamps that are missing
 * from the time series data with `null` values.
 */
export function getTimeSeriesValues(series: TimeSeries, timeScale: TimeScale) {
  let timestamp = timeScale.startMs;

  const values = series.values;
  const processedValues: TimeSeriesValueTuple[] = [];

  for (const valueTuple of values) {
    // Fill in values up to the current series value timestamp with nulls
    while (timestamp < valueTuple[0]) {
      processedValues.push([timestamp, null]);
      timestamp += timeScale.stepMs;
    }

    // Now add the current value since timestamp should match
    processedValues.push([timestamp, valueTuple[1]]);
    timestamp += timeScale.stepMs;
  }

  // Add null values at the end of the series if necessary
  while (timestamp <= timeScale.endMs) {
    processedValues.push([timestamp, null]);
    timestamp += timeScale.stepMs;
  }

  return processedValues;
}

/**
 * Given a TimeSeries from a query and a common time scale (see `getCommonTimeScale`),
 * gets the values for the y axis of a graph, filling in any timestamps that are
 * missing from the time series data with `null` values.
 */
export function getYValues(series: TimeSeries, timeScale: TimeScale): Array<number | null> {
  let timestamp = timeScale.startMs;

  const yValues: Array<number | null> = [];
  for (const valueTuple of series.values) {
    // Fill in values up to the current series value timestamp with nulls
    while (timestamp < valueTuple[0]) {
      yValues.push(null);
      timestamp += timeScale.stepMs;
    }

    // Now add the current value since timestamp should match
    yValues.push(valueTuple[1]);
    timestamp += timeScale.stepMs;
  }

  // Add null values at the end of the series if necessary
  while (timestamp <= timeScale.endMs) {
    yValues.push(null);
    timestamp += timeScale.stepMs;
  }

  return yValues;
}

/**
 * Given a list of running queries, calculates a common time scale for use on
 * the x axis (i.e. start/end dates and a step that is divisible into all of
 * the queries' steps).
 */
export function getCommonTimeScale(
  seriesData: Array<TimeSeriesData | Pick<TimeSeries, 'values'> | undefined>
): TimeScale | undefined {
  let timeRange: AbsoluteTimeRange | undefined = undefined;
  const steps: number[] = [];

  for (const data of seriesData) {
    if (data && 'timeRange' in data) {
      if (data === undefined || data.timeRange === undefined || data.stepMs === undefined) continue;

      // Keep track of query steps so we can calculate a common one for the graph
      steps.push(data.stepMs);

      // If we don't have an overall time range yet, just start with this one
      if (timeRange === undefined) {
        timeRange = data.timeRange;
        continue;
      }

      // Otherwise, see if this query has a start or end outside of the current
      // time range
      if (data.timeRange.start < timeRange.start) {
        timeRange.start = data.timeRange.start;
      }
      if (data.timeRange.end > timeRange.end) {
        timeRange.end = data.timeRange.end;
      }
    } else if (data && 'values' in data) {
      for (let i = 0; i < data.values.length; i++) {
        const values = data.values[i];
        if (values === undefined) {
          continue;
        }

        const [timestamp] = values;
        const start = new Date(timestamp);
        const end = new Date(timestamp);

        if (timeRange === undefined) {
          timeRange = {
            start,
            end,
          };
          continue;
        }

        if (start < timeRange.start) {
          timeRange.start = start;
        }

        if (end > timeRange.end) {
          timeRange.end = end;
        }
      }

      if (timeRange === undefined) return undefined;

      const startMs = timeRange.start.valueOf();
      const endMs = timeRange.end.valueOf();
      const rangeMs = endMs - startMs;

      return { startMs, endMs, rangeMs, stepMs: MIN_STEP_INTERVAL_MS };
    }
  }

  if (timeRange === undefined) return undefined;

  // Use the greatest common divisor of all step values as the overall step
  // for the x axis (or if only one query, just use that query's step value)
  let stepMs: number;
  if (steps.length === 1) {
    stepMs = steps[0] as number;
  } else {
    const calculatedStepMs = gcd(...steps);
    stepMs = calculatedStepMs < MIN_STEP_INTERVAL_MS ? MIN_STEP_INTERVAL_MS : calculatedStepMs;
  }

  const startMs = timeRange.start.valueOf();
  const endMs = timeRange.end.valueOf();
  const rangeMs = endMs - startMs;

  return { startMs, endMs, stepMs, rangeMs };
}
