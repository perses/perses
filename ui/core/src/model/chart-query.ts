import { DataFrame } from './data-frame';
import { Definition, JsonObject } from './definitions';
import { AnyPluginDefinition, AnyPluginImplementation } from './plugins';
import { ResourceSelector } from './resource';

export interface ChartQueryDefinition<
  Kind extends string,
  Options extends JsonObject
> extends Definition<Kind, Options> {
  datasource?: ResourceSelector;
}

export interface ChartQueryPlugin<
  Kind extends string,
  Options extends JsonObject
> {
  useChartQuery: UseChartQueryHook<Kind, Options>;
}

export type UseChartQueryHook<
  Kind extends string,
  Options extends JsonObject
> = (definition: ChartQueryDefinition<Kind, Options>) => {
  data: DataFrame[];
  loading: boolean;
  error?: Error;
};

export type AnyChartQueryDefinition = AnyPluginDefinition<'ChartQuery'>;

export type AnyChartQueryPlugin = AnyPluginImplementation<'ChartQuery'>;
