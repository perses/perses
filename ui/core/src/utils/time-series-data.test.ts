// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TimeScale, TimeSeries, TimeSeriesData } from '../model';
import { getXValues, getYValues, getCommonTimeScale } from './time-series-data';

function getExpectedIntervals(timeScale: TimeScale): number {
  return Math.floor((timeScale.endMs - timeScale.startMs) / timeScale.stepMs) + 1;
}

type GenerateMockTimeSeriesOpts = {
  /**
   * Name of the series
   */
  name: string;

  /**
   * Number of entries to generate.
   */
  count: number;

  /**
   * Timestamp to starget generating values.
   */
  startMs: number;

  /**
   * Step to use when generating additional values.
   */
  stepMs: number;
};

function generateMockTimeSeries({ name, count, startMs, stepMs }: GenerateMockTimeSeriesOpts): TimeSeries {
  return {
    name: name,
    values: [...Array(count)].map((_, i) => {
      return [startMs + stepMs * i, i];
    }),
  };
}

describe('getXValues', () => {
  test('returns timestamp values for the x axis', () => {
    const timeScale: TimeScale = {
      startMs: 1646164800000,
      endMs: 1646251200000,
      stepMs: 3600000,
    };
    const xValues = getXValues(timeScale);

    // Start and end of time scale are retained.
    expect(xValues[0]).toEqual(timeScale.startMs);
    expect(xValues.at(-1)).toEqual(timeScale.endMs);

    // The expected number of intervals are included in the results
    expect(xValues).toHaveLength(getExpectedIntervals(timeScale));

    expect(xValues).toMatchSnapshot();
  });
});

describe('getYValues', () => {
  test('uses the specified values if they have no missing timestamps', () => {
    const timeScale: TimeScale = {
      startMs: 1646164800000,
      endMs: 1646251200000,
      stepMs: 3600000,
    };

    const series = generateMockTimeSeries({
      name: 'my series',
      count: getExpectedIntervals(timeScale),
      startMs: timeScale.startMs,
      stepMs: timeScale.stepMs,
    });

    const yValues = getYValues(series, timeScale);
    yValues.forEach((entry) => {
      expect(entry).not.toBeNull();
    });
    expect(yValues).toMatchSnapshot();
  });

  test('fills in null values when there are no values', () => {
    const series: TimeSeries = {
      name: 'my series',
      values: [],
    };
    const timeScale: TimeScale = {
      startMs: 1646164800000,
      endMs: 1646251200000,
      stepMs: 3600000,
    };
    const yValues = getYValues(series, timeScale);
    expect(yValues).toMatchSnapshot();
  });

  test('fills in null values when missing some values at the end', () => {
    const timeScale: TimeScale = {
      startMs: 1646164800000,
      endMs: 1646251200000,
      stepMs: 3600000,
    };
    const expectedIntervals = getExpectedIntervals(timeScale);
    const expectedNulls = Math.ceil(expectedIntervals / 2);

    const series = generateMockTimeSeries({
      name: 'my series',
      count: expectedIntervals - expectedNulls + 1,
      startMs: timeScale.startMs,
      stepMs: timeScale.stepMs,
    });

    const yValues = getYValues(series, timeScale);
    expect(yValues).toMatchSnapshot();
  });

  test('fills in null values when missing some values at the beginning', () => {
    const timeScale: TimeScale = {
      startMs: 1646164800000,
      endMs: 1646251200000,
      stepMs: 3600000,
    };
    const expectedIntervals = getExpectedIntervals(timeScale);
    const expectedNulls = Math.ceil(expectedIntervals / 2);

    const series = generateMockTimeSeries({
      name: 'my series',
      count: expectedIntervals - expectedNulls + 1,
      startMs: timeScale.startMs + Math.floor((timeScale.endMs - timeScale.startMs) / 2),
      stepMs: timeScale.stepMs,
    });

    const yValues = getYValues(series, timeScale);
    expect(yValues).toMatchSnapshot();
  });

  test('fills in null values when missing some values in the middle', () => {
    const timeScale: TimeScale = {
      startMs: 1646164800000,
      endMs: 1646251200000,
      stepMs: 3600000,
    };
    const expectedIntervals = getExpectedIntervals(timeScale);
    const expectedNulls = Math.ceil(expectedIntervals / 2);

    const series = generateMockTimeSeries({
      name: 'my series',
      count: expectedIntervals - expectedNulls + 1,
      startMs: timeScale.startMs + Math.floor((timeScale.endMs - timeScale.startMs) / 4),
      stepMs: timeScale.stepMs,
    });

    const yValues = getYValues(series, timeScale);
    expect(yValues).toMatchSnapshot();
  });
});

describe('getCommonTimeScale', () => {
  const undefinedSeriesData: Array<{
    name: string;
    seriesData: Array<TimeSeriesData | undefined>;
  }> = [
    {
      name: 'when series data is empty',
      seriesData: [],
    },
    {
      name: 'when series is missing step',
      seriesData: [
        {
          timeRange: {
            start: new Date('Tue Mar 28 2023 16:47:00 GMT-0700 (Pacific Daylight Time)'),
            end: new Date('Tue Mar 28 2023 04:47:00 GMT-0700 (Pacific Daylight Time)'),
          },
          series: [],
        },
      ],
    },
    {
      name: 'when series is missing time range',
      seriesData: [
        {
          stepMs: 6000,
          series: [],
        },
      ],
    },
  ];

  test.each(undefinedSeriesData)('returns undefined $name', ({ seriesData }) => {
    expect(getCommonTimeScale(seriesData)).toBeUndefined;
  });

  test('uses the specified time scale when a single series is available', () => {
    const singleTimeSeries: TimeSeriesData = {
      timeRange: {
        start: new Date('Tue Mar 28 2023 16:47:00 GMT-0700 (Pacific Daylight Time)'),
        end: new Date('Tue Mar 28 2023 04:47:00 GMT-0700 (Pacific Daylight Time)'),
      },
      stepMs: 3600,
      series: [],
    };
    const seriesData: Array<TimeSeriesData | undefined> = [singleTimeSeries];
    const timeScale = getCommonTimeScale(seriesData);

    expect(timeScale).toEqual({
      startMs: singleTimeSeries.timeRange?.start.getTime(),
      endMs: singleTimeSeries.timeRange?.end.getTime(),
      stepMs: singleTimeSeries.stepMs,
    });
  });

  test('uses the min/max dates when multiple ranges available', () => {
    const minTime = new Date('Tue Mar 21 2023 17:00:00 GMT-0700 (Pacific Daylight Time)');
    const maxTime = new Date('Tue Mar 28 2023 17:00:00 GMT-0700 (Pacific Daylight Time)');

    const seriesData: Array<TimeSeriesData | undefined> = [
      {
        timeRange: {
          start: minTime,
          end: new Date('Tue Mar 28 2023 04:47:00 GMT-0700 (Pacific Daylight Time)'),
        },
        stepMs: 3600,
        series: [],
      },
      {
        timeRange: {
          start: new Date('Tue Mar 28 2023 16:47:00 GMT-0700 (Pacific Daylight Time)'),
          end: new Date('Tue Mar 28 2023 17:00:00 GMT-0700 (Pacific Daylight Time)'),
        },
        stepMs: 3600,
        series: [],
      },
      {
        timeRange: {
          start: new Date('Tue Mar 28 2023 16:47:00 GMT-0700 (Pacific Daylight Time)'),
          end: maxTime,
        },
        stepMs: 3600,
        series: [],
      },
    ];
    const timeScale = getCommonTimeScale(seriesData);
    expect(timeScale).toEqual({
      startMs: minTime.getTime(),
      endMs: maxTime.getTime(),
      stepMs: 3600,
    });
  });

  test('uses the greatest common denominator when multiple steps are available', () => {
    const timeRange = {
      start: new Date('Tue Mar 28 2023 16:47:00 GMT-0700 (Pacific Daylight Time)'),
      end: new Date('Tue Mar 28 2023 17:00:00 GMT-0700 (Pacific Daylight Time)'),
    };
    const steps = [2400, 3600, 18000];

    const seriesData: Array<TimeSeriesData | undefined> = [
      {
        timeRange: timeRange,
        stepMs: steps[0],
        series: [],
      },
      {
        timeRange: timeRange,
        stepMs: steps[1],
        series: [],
      },
      {
        timeRange: timeRange,
        stepMs: steps[2],
        series: [],
      },
    ];
    const timeScale = getCommonTimeScale(seriesData);
    expect(timeScale).toEqual({
      startMs: timeRange.start.getTime(),
      endMs: timeRange.end.getTime(),
      stepMs: 1200,
    });
  });
});
