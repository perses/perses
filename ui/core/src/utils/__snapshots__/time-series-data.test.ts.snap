// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getXValues returns timestamp values for the x axis 1`] = `
Array [
  1646164800000,
  1646168400000,
  1646172000000,
  1646175600000,
  1646179200000,
  1646182800000,
  1646186400000,
  1646190000000,
  1646193600000,
  1646197200000,
  1646200800000,
  1646204400000,
  1646208000000,
  1646211600000,
  1646215200000,
  1646218800000,
  1646222400000,
  1646226000000,
  1646229600000,
  1646233200000,
  1646236800000,
  1646240400000,
  1646244000000,
  1646247600000,
  1646251200000,
]
`;

exports[`getYValues fills in null values when missing some values at the beginning 1`] = `
Array [
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
]
`;

exports[`getYValues fills in null values when missing some values at the end 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
]
`;

exports[`getYValues fills in null values when missing some values in the middle 1`] = `
Array [
  null,
  null,
  null,
  null,
  null,
  null,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  null,
  null,
  null,
  null,
  null,
  null,
]
`;

exports[`getYValues fills in null values when there are no values 1`] = `
Array [
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
]
`;

exports[`getYValues uses the specified values if they have no missing timestamps 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
]
`;
